File: 3313-maximum-strength-of-k-disjoint-subarrays/maximum-strength-of-k-disjoint-subarrays.py
class Solution:
    def maximumStrength(self, nums: List[int], k: int) -> int:
        """
        dp[i][j]: max strength i subarrays, up to j index 
        """
        n = len(nums)
        dp = [[0]*(n+1) for _ in range(k+1)]

        for i in range(1,k+1):
            multiplier = k+1-i if i%2==1 else i-k-1
            maxsum,cur = -inf,-inf
            for j in range(i-1,n):
                cur = max(cur+nums[j]*multiplier, dp[i-1][j]+nums[j]*multiplier)
                maxsum = max(maxsum, cur)
                dp[i][j+1] = maxsum

        return dp[k][n]

File: 3313-maximum-strength-of-k-disjoint-subarrays/maximum-strength-of-k-disjoint-subarrays.java
class Solution {
    public long maximumStrength(int[] nums, int k) {
        int n = nums.length;
        long[][] dp = new long[k+1][n+1];
        for (long[] row : dp)
            Arrays.fill(row, 0l);
        
        for (int i=1;i<=k;i++) {
            long maxsum = Long.MIN_VALUE/2;
            long curr = Long.MIN_VALUE/2;
            long multiplier = i%2==1 ? k+1-i : i-k-1;
            for (int j=i-1;j<n;j++) {
                curr = Math.max(curr+nums[j]*multiplier, dp[i-1][j]+nums[j]*multiplier);
                maxsum = Math.max(maxsum, curr);
                dp[i][j+1] = maxsum;
            }
        }

        return dp[k][n];
    }
}